File: Depth_first_search_algorithm/Bài_toán_12.py
def dfs_multigraph(adj, u, visited=None, result=None):
    if visited is None:
        visited = set()
    if result is None:
        result = []
    visited.add(u)
    result.append(u)
    for v in adj[u]:
        if v not in visited:
            dfs_multigraph(adj, v, visited, result)
    return result

graph = {
    0: [1, 1, 2],
    1: [0, 0, 2],
    2: [0, 1, 2],  # self-loop at 2
    3: []
}

print("DFS traversal:", dfs_multigraph(graph, 0))

File: Depth_first_search_algorithm/Bài_toán_12.cpp
#include <iostream>
#include <vector>
using namespace std;

void dfs_multigraph(int u, const vector<vector<int>>& adj, vector<bool>& visited) {
    visited[u] = true;
    cout << u << " ";
    for (int v : adj[u]) {
        if (!visited[v])
            dfs_multigraph(v, adj, visited);
    }
}

int main() {
    vector<vector<int>> graph = {
        {1, 1, 2},
        {0, 0, 2},
        {0, 1, 2},
        {}
    };

    int n = graph.size();
    vector<bool> visited(n, false);
    cout << "DFS traversal: ";
    dfs_multigraph(0, graph, visited);
    cout << endl;

    return 0;
}
